art/attacks/attack.py

```diff
@@ -163,28 +163,62 @@ class PoisoningAttack(Attack):
     Abstract base class for poisoning attack classes
     """
 
-    def __init__(self, classifier) -> None:
+    def __init__(self, classifier: Optional[Classifier]) -> None:
+        """
+        :param classifier: A trained classifier (or none if no classifier is needed)
+        """
+        super().__init__(classifier)
+
+    @abc.abstractmethod
+    def poison(self, x: np.ndarray, y=Optional[np.ndarray], **kwargs) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Generate poisoning examples and return them as an array. This method should be overridden by all concrete
+        poisoning attack implementations.
+
+        :param x: An array with the original inputs to be attacked.
+        :param y:  Target labels for `x`. Untargeted attacks set this value to None.
+        :return: An tuple holding the (poisoning examples, poisoning labels).
+        """
+        raise NotImplementedError
+
+
+class PoisoningAttackTransformer(PoisoningAttack):
+    """
+    Abstract base class for poisoning attack classes that affect the classifier.
+    These attacks have an additional method, `poison_estimator`, that returns the poisoned classifier.
+    """
+    ClassifierType = TypeVar('ClassifierType', bound=Classifier)
```

Comment:
Could we move this to a place in this file from where we can reuse it in the return type of `ExtractionAttack.extract`?

Reply:
Not sure if this will ruin things with scoping in Python, but I can replicate this for extraction attacks as well.

Reply:
Actually I'm still unsure how to do this given there is no classifier type defined in an `__init__` function for extraction attacks, and estimator types are handled within the general Attack class